src/contracts/Qearn.h

```diff
@@ -0,0 +1,866 @@
+        {
+            locals.INITIALIZE_ROUNDINFO._Epoch_Bonus_Amount = locals.current_balance - locals.pre_epoch_balance;
+        }
+        locals.INITIALIZE_ROUNDINFO._Total_Locked_Amount = 0;
+
+        state._InitialRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+        state._CurrentRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+
+	_
+
+    struct END_EPOCH_locals 
+    {
+        HistoryInfo INITIALIZE_HISTORY;
+        LockInfo INITIALIZE_USER;
+        RoundInfo INITIALIZE_ROUNDINFO;
+        EpochIndexInfo tmpEpochIndex;
+
+        uint64 _reward_percent;
+        uint64 _reward_amount;
+        uint64 _burn_amount;
+        uint32 locked_epoch;
+        uint32 startEpoch;
+        uint32 _t;
+        sint32 st;
+        sint32 en;
+        uint32 empty_cnt;
+        uint32 endIndex;
+
+    };
+
+	END_EPOCH_WITH_LOCALS
+
+        state._EarlyUnlocked_cnt = 0;
+        state._FullyUnlocked_cnt = 0;
+        locals.locked_epoch = qpi.epoch() - 52;
+        locals.endIndex = state.EpochIndex.get(locals.locked_epoch).endIndex;
+        
+        locals._burn_amount = state._CurrentRoundInfo.get(locals.locked_epoch)._Epoch_Bonus_Amount;
+        
+        locals._reward_percent = QPI::div(state._CurrentRoundInfo.get(locals.locked_epoch)._Epoch_Bonus_Amount * 10000000ULL, state._CurrentRoundInfo.get(locals.locked_epoch)._Total_Locked_Amount);
+
+        for(locals._t = state.EpochIndex.get(locals.locked_epoch).startIndex; locals._t < locals.endIndex; locals._t++) 
+        {
+            if(state.Locker.get(locals._t)._Locked_Amount == 0) 
+            {
+                continue;
+            }
+
+            locals._reward_amount = QPI::div(state.Locker.get(locals._t)._Locked_Amount * locals._reward_percent, 10000000ULL);
+            qpi.transfer(state.Locker.get(locals._t).ID, locals._reward_amount + state.Locker.get(locals._t)._Locked_Amount);
+
+            if(state._FullyUnlocked_cnt < QEARN_MAX_USERS) 
+            {
+
+                locals.INITIALIZE_HISTORY._Unlocked_ID = state.Locker.get(locals._t).ID;
+                locals.INITIALIZE_HISTORY._Unlocked_Amount = state.Locker.get(locals._t)._Locked_Amount;
+                locals.INITIALIZE_HISTORY._Rewarded_Amount = locals._reward_amount;
+
+                state.FullyUnlocker.set(state._FullyUnlocked_cnt, locals.INITIALIZE_HISTORY);
+
+                state._FullyUnlocked_cnt++;
+            }
+
+            locals.INITIALIZE_USER.ID = NULL_ID;
+            locals.INITIALIZE_USER._Locked_Amount = 0;
+            locals.INITIALIZE_USER._Locked_Epoch = 0;
+
+            state.Locker.set(locals._t, locals.INITIALIZE_USER);
+
+            locals._burn_amount -= locals._reward_amount;
+        }
+
+        locals.st = 0;
+
+        for(locals._t = 0; locals._t < QEARN_MAX_LOCKS; locals._t++)
+        {
+            if(state.Locker.get(locals._t)._Locked_Epoch) 
+            {
+                locals.startEpoch = state.Locker.get(locals._t)._Locked_Epoch;
```

Comment:
After a lot of automated testing with random data, I have discovered another bug:
If early unlocking happens, all items of an epoch may be removed from the Locker array. Then startEpoch is behind this epoch. The consequences: epochIndex is not updated for epoch with all unlocked early, which can later lead to a wrong payout at END_EPOCH of this epoch (payout of locker entry not belonging to the epoch).

I have implemented a fix for making sure the epochIndex is consistent, also after payout. I test it at the moment and will provide it soon.